transport/http/jsonrpc/server.go

```diff
@@ -0,0 +1,205 @@
+	ecm EndpointCodecMap,
+	options ...ServerOption,
+) *Server {
+	s := &Server{
+		ecm:          ecm,
+		errorEncoder: DefaultErrorEncoder,
+		logger:       log.NewNopLogger(),
+	}
+	for _, option := range options {
+		option(s)
+	}
+	return s
+}
+
+// ServerOption sets an optional parameter for servers.
+type ServerOption func(*Server)
+
+// ServerBefore functions are executed on the HTTP request object before the
+func ServerBefore(before ...httptransport.RequestFunc) ServerOption {
+	// request is decoded.
+	return func(s *Server) { s.before = append(s.before, before...) }
+}
+
+// ServerAfter functions are executed on the HTTP response writer after the
+// endpoint is invoked, but before anything is written to the client.
+func ServerAfter(after ...httptransport.ServerResponseFunc) ServerOption {
+	return func(s *Server) { s.after = append(s.after, after...) }
+}
+
+// ServerErrorEncoder is used to encode errors to the http.ResponseWriter
+// whenever they're encountered in the processing of a request. Clients can
+// use this to provide custom error formatting and response codes. By default,
+// errors will be written with the DefaultErrorEncoder.
+func ServerErrorEncoder(ee httptransport.ErrorEncoder) ServerOption {
+	return func(s *Server) { s.errorEncoder = ee }
+}
+
+// ServerErrorLogger is used to log non-terminal errors. By default, no errors
+// are logged. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ServerErrorEncoder or ServerFinalizer, both of which have access to
+// the context.
+func ServerErrorLogger(logger log.Logger) ServerOption {
+	return func(s *Server) { s.logger = logger }
+}
+
+// ServerFinalizer is executed at the end of every HTTP request.
+// By default, no finalizer is registered.
+func ServerFinalizer(f httptransport.ServerFinalizerFunc) ServerOption {
+	return func(s *Server) { s.finalizer = f }
+}
+
+// ServeHTTP implements http.Handler.
+func (s Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		io.WriteString(w, "405 must POST\n")
+		return
+	}
+	ctx := r.Context()
+
+	if s.finalizer != nil {
+		iw := &interceptingWriter{w, http.StatusOK}
+		defer func() { s.finalizer(ctx, iw.code, r) }()
+		w = iw
+	}
+
+	for _, f := range s.before {
+		ctx = f(ctx, r)
+	}
+
+	// Decode the body into an  object
+	var req Request
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		rpcerr := parseError("JSON could not be decoded: " + err.Error())
+		s.logger.Log("err", rpcerr)
+		s.errorEncoder(ctx, rpcerr, w)
+		return
+	}
+
+	// Get the endpoint and codecs from the map using the method
+	// defined in the JSON  object
+	ecm, ok := s.ecm[req.Method]
+	if !ok {
+		err := methodNotFoundError(fmt.Sprintf("Method %s was not found.", req.Method))
```

Comment:
Oh, I see now how you're using the error types. Hmm, interesting. Disregard my previous comment.

Reply:
I confused myself with this at times. Glad it makes sense to someone else!